src/nxt_cgroup.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) Andrew Clayton
+ * Copyright (C) F5, Inc.
+ */
+
+#include <nxt_main.h>
+#include <nxt_process.h>
+#include <nxt_fs.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <limits.h>
+
+
+static nxt_int_t nxt_make_cgpaths(const nxt_process_t *process, char *cgpath,
+    size_t cgpath_len, char *cgroot, size_t cgroot_len);
+static nxt_int_t nxt_cgroup_create(const nxt_task_t *task,
+    nxt_process_t *process);
+nxt_int_t nxt_cgroup_proc_add(const nxt_task_t *task, nxt_process_t *process);
+void nxt_cgroup_cleanup(const nxt_process_t *process);
+
+
+static nxt_int_t
+nxt_make_cgpaths(const nxt_process_t *process, char *cgpath, size_t cgpath_len,
+    char *cgroot, size_t cgroot_len)
+{
+    snprintf(cgpath, cgpath_len, NXT_CGROUP_ROOT "/%s",
+             process->isolation.cgroup.name);
+    if (cgroot != NULL) {
+        snprintf(cgroot, cgroot_len, "%s", NXT_CGROUP_ROOT);
+    }
+
+    return NXT_OK;
+}
+
+
+static nxt_int_t
+nxt_cgroup_create(const nxt_task_t *task, nxt_process_t *process)
+{
+    char    cgroot[PATH_MAX];
+    char    cgpath[PATH_MAX];
+
+    if (task->thread->runtime->type != NXT_PROCESS_MAIN
+        || nxt_process_type(process) != NXT_PROCESS_PROTOTYPE) {
+        return NXT_OK;
+    }
+
+    if (process->isolation.cgroup.name == NULL) {
+        return NXT_OK;
+    }
+
+    nxt_make_cgpaths(process, cgpath, sizeof(cgpath), cgroot, sizeof(cgroot));
+    nxt_fs_mkdir_all((const u_char *)cgpath, 0777);
+
+    return NXT_OK;
+}
+
+
+nxt_int_t
+nxt_cgroup_proc_add(const nxt_task_t *task, nxt_process_t *process)
+{
+    FILE    *fp;
+    char    cgpath[PATH_MAX];
+    char    cgprocs[PATH_MAX];
+
+    nxt_cgroup_create(task, process);
+
+    if (nxt_process_type(process) != NXT_PROCESS_PROTOTYPE) {
+        return NXT_OK;
+    }
+
+    if (process->isolation.cgroup.name == NULL) {
+        return NXT_OK;
+    }
+
+    nxt_make_cgpaths(process, cgpath, sizeof(cgpath), NULL, 0);
+    snprintf(cgprocs, sizeof(cgprocs) - strlen(cgpath) - 14, "%s/cgroup.procs",
```

Comment:
Don't we know the length already?  Can we skip a `strlen(3)` call?

Reply:
No, we don't know what length of string was put into cgpath, but I'm going to be changing how that's done anyway and can get the length then.

Reply:
nxt_make_cgpaths() is no more...